File: assignments/assignment-1-solution/src/CrazyEights/Game/TurnAction.cs
using CrazyEights.CardDeck;
using CrazyEights.Cards;
using CrazyEights.Domain;
using CrazyEights.Player;

namespace CrazyEights.Game;

/// <summary>
/// Implements turn execution steps and game UI prompts.
/// </summary>
public static class TurnAction
{
    public static void ShowTurn(GameContext gameContext, TurnContext turnContext)
    {
        var turnCount = gameContext.TurnNumber;
        var players = gameContext.Players;

        GameConsole.WriteLine($"----- TURN {turnCount} -----");
        ShowTopCard(turnContext.TopCard, turnContext.CurrentSuit);
        ShowRemainingDeckCount(gameContext.Deck);
        ShowCardCounts(players);

        ShowMessage($"{turnContext.CurrentPlayer.Name}'s turn".ToUpper(), true);
    }

    public static void StartTurn(GameContext gameContext, TurnContext turnContext)
    {
        var player = turnContext.CurrentPlayer;

        if (player.ShowHand)
            ShowAllCardsInHand(turnContext);

        SelectCard(gameContext, turnContext);
    }

    private static ICard DrawCard(GameContext gameContext, TurnContext turnContext)
    {
        var player = turnContext.CurrentPlayer;
        var name = player.Name;
        ShowMessage($"{name} has no playable cards. Drawing one card...", true);

        var drawnCard = gameContext.Deck.DrawCard();
        player.AddCard(drawnCard);

        var cardDescription = player.ShowHand ? drawnCard.GetDescription() : "a card";
        ShowMessage($"{name} drew {cardDescription}");

        var canPlayDrawnCard = PlayableCardsSelector.CanPlayCard(
            drawnCard,
            turnContext.CurrentSuit,
            turnContext.CurrentRank);
        if (canPlayDrawnCard.CanPlay && player.WillPlayDrawnCard(turnContext, drawnCard)) return drawnCard;

        GameConsole.ReadLine("Press Enter to continue...");
        return UnselectedCard.Instance;
    }

    private static void SelectCard(GameContext gameContext, TurnContext turnContext)
    {
        var currentPlayer = turnContext.CurrentPlayer;
        var selectedCard = currentPlayer.SelectCard(turnContext);

        if (!selectedCard.IsSelectable)
        {
            var drawnCard = DrawCard(gameContext, turnContext);
            if (!drawnCard.IsSelectable) return;
            selectedCard = drawnCard;
        }

        currentPlayer.RemoveCard(selectedCard);
        gameContext.DiscardPile.AddCard(selectedCard);
        var name = turnContext.CurrentPlayer.Name;
        ShowMessage($"{name} selected {selectedCard.GetDescription()}");

        if (Rank.IsWildcardRank(selectedCard.Rank)) ChooseSuit(gameContext, turnContext);
    }

    private static void ChooseSuit(GameContext gameContext, TurnContext turnContext)
    {
        var currentPlayer = turnContext.CurrentPlayer;
        var chosenSuit = currentPlayer.SelectSuit(gameContext, turnContext);
        var name = turnContext.CurrentPlayer.Name;

        var discardPile = gameContext.DiscardPile;
        if (discardPile.ActiveSuit != chosenSuit)
        {
            discardPile.OverrideTopCardSuit(chosenSuit);
            ShowMessage($"{name} changed suit to {chosenSuit}");
            return;
        }

        ShowMessage($"{name} left the suit as {chosenSuit}");
    }

    private static void ShowMessage(string action, bool showBlankLineBefore = false)
    {
        if (showBlankLineBefore) GameConsole.WriteLine();
        GameConsole.WriteLine($"** {action}");
    }

    private static void ShowAllCardsInHand(TurnContext turnContext)
    {
        var name = turnContext.CurrentPlayer.Name;
        var hand = turnContext.CurrentPlayer.PeekHand();

        Console.WriteLine();
        GameConsole.WriteLine($"{name}'s hand");

        foreach (var card in hand) GameConsole.WriteLine($"  - {card.GetDescription()}");
    }

    private static void ShowRemainingDeckCount(Deck deck)
    {
        var cardCount = deck.CardCount;
        GameConsole.WriteLine($"Deck remaining: {cardCount} {GetPluralCardLabel(cardCount)}");
    }

    public static void ShowTopCard(ICard topCard, SuitType currentSuit, string prefix = "Top discard: ")
    {
        var isSuitChanged = topCard.Suit != currentSuit;
        var suitToMatch = isSuitChanged
            ? $" (Suit to match: {currentSuit})"
            : "";

        GameConsole.WriteLine($"{prefix}{topCard.GetDescription()}{suitToMatch}");
    }

    private static void ShowCardCounts(Players players)
    {
        for (var i = 0; i < players.List.Count; i++)
        {
            if (i > 0) GameConsole.Write(" | ");
            var player = players.List[i];
            var name = player.Name;
            var cardCount = player.CardCount;
            GameConsole.Write($"{name}: {cardCount} {GetPluralCardLabel(cardCount)}");
        }

        GameConsole.WriteLine();
    }

    private static string GetPluralCardLabel(int count)
    {
        return count == 1 ? "card" : "cards";
    }
}

File: assignments/assignment-1-solution/src-java/src/main/java/edu/kennesaw/crazy8s/game/TurnAction.java
package edu.kennesaw.crazy8s.game;

import edu.kennesaw.crazy8s.carddeck.Deck;
import edu.kennesaw.crazy8s.carddeck.DiscardPile;
import edu.kennesaw.crazy8s.cards.Card;
import edu.kennesaw.crazy8s.cards.UnselectedCard;
import edu.kennesaw.crazy8s.domain.IsPlayable;
import edu.kennesaw.crazy8s.domain.PlayableCard;
import edu.kennesaw.crazy8s.domain.PlayableCardsSelector;
import edu.kennesaw.crazy8s.domain.Rank;
import edu.kennesaw.crazy8s.domain.SuitType;
import edu.kennesaw.crazy8s.player.Player;
import edu.kennesaw.crazy8s.player.Players;
import java.util.List;

/**
 * Implements turn execution steps and game UI prompts.
 */
public final class TurnAction {
    private TurnAction() {
    }

    public static void showTurn(GameContext gameContext, TurnContext turnContext) {
        int turnCount = gameContext.getTurnNumber();
        Players players = gameContext.getPlayers();

        GameConsole.writeLine(String.format("----- TURN %d -----", turnCount));
        showTopCard(turnContext.getTopCard(), turnContext.getCurrentSuit());
        showRemainingDeckCount(gameContext.getDeck());
        showCardCounts(players);

        showMessage(turnContext.getCurrentPlayer().getName().toUpperCase() + "'s turn", true);
    }

    public static void startTurn(GameContext gameContext, TurnContext turnContext) {
        Player player = turnContext.getCurrentPlayer();

        if (player.isShowHand()) {
            showAllCardsInHand(turnContext);
        }

        selectCard(gameContext, turnContext);
    }

    private static Card drawCard(GameContext gameContext, TurnContext turnContext) {
        Player player = turnContext.getCurrentPlayer();
        String name = player.getName();
        showMessage(name + " has no playable cards. Drawing one card...", true);

        Card drawnCard = gameContext.getDeck().drawCard();
        player.addCard(drawnCard);

        String cardDescription = player.isShowHand() ? drawnCard.getDescription() : "a card";
        showMessage(name + " drew " + cardDescription);

        IsPlayable canPlayDrawnCard = PlayableCardsSelector.canPlayCard(
                drawnCard,
                turnContext.getCurrentSuit(),
                turnContext.getCurrentRank());
        if (canPlayDrawnCard.canPlay() && player.willPlayDrawnCard(turnContext, drawnCard)) {
            return drawnCard;
        }

        GameConsole.readLine("Press Enter to continue...");
        return UnselectedCard.getInstance();
    }

    private static void selectCard(GameContext gameContext, TurnContext turnContext) {
        Player currentPlayer = turnContext.getCurrentPlayer();
        Card selectedCard = currentPlayer.selectCard(turnContext);

        if (!selectedCard.isSelectable()) {
            Card drawnCard = drawCard(gameContext, turnContext);
            if (!drawnCard.isSelectable()) {
                return;
            }
            selectedCard = drawnCard;
        }

        currentPlayer.removeCard(selectedCard);
        gameContext.getDiscardPile().addCard(selectedCard);
        String name = turnContext.getCurrentPlayer().getName();
        showMessage(name + " selected " + selectedCard.getDescription());

        if (Rank.isWildcardRank(selectedCard.getRank())) {
            chooseSuit(gameContext, turnContext);
        }
    }

    private static void chooseSuit(GameContext gameContext, TurnContext turnContext) {
        Player currentPlayer = turnContext.getCurrentPlayer();
        SuitType chosenSuit = currentPlayer.selectSuit(gameContext, turnContext);
        String name = turnContext.getCurrentPlayer().getName();

        DiscardPile discardPile = gameContext.getDiscardPile();
        if (discardPile.getActiveSuit() != chosenSuit) {
            discardPile.overrideTopCardSuit(chosenSuit);
            showMessage(name + " changed suit to " + chosenSuit);
            return;
        }

        showMessage(name + " left the suit as " + chosenSuit);
    }

    private static void showMessage(String action) {
        showMessage(action, false);
    }

    private static void showMessage(String action, boolean showBlankLineBefore) {
        if (showBlankLineBefore) {
            GameConsole.writeLine();
        }
        GameConsole.writeLine("** " + action);
    }

    private static void showAllCardsInHand(TurnContext turnContext) {
        String name = turnContext.getCurrentPlayer().getName();
        List<Card> hand = turnContext.getCurrentPlayer().peekHand();

        GameConsole.writeLine();
        GameConsole.writeLine(name + "'s hand");

        for (Card card : hand) {
            GameConsole.writeLine("  - " + card.getDescription());
        }
    }

    private static void showRemainingDeckCount(Deck deck) {
        int cardCount = deck.getCardCount();
        GameConsole.writeLine("Deck remaining: " + cardCount + " " + getPluralCardLabel(cardCount));
    }

    public static void showTopCard(Card topCard, SuitType currentSuit) {
        showTopCard(topCard, currentSuit, "Top discard: ");
    }

    public static void showTopCard(Card topCard, SuitType currentSuit, String prefix) {
        boolean isSuitChanged = topCard.getSuit() != currentSuit;
        String suitToMatch = isSuitChanged
                ? " (Suit to match: " + currentSuit + ")"
                : "";

        GameConsole.writeLine(prefix + topCard.getDescription() + suitToMatch);
    }

    private static void showCardCounts(Players players) {
        List<Player> playerList = players.getList();
        for (int i = 0; i < playerList.size(); i++) {
            if (i > 0) {
                GameConsole.write(" | ");
            }
            Player player = playerList.get(i);
            String name = player.getName();
            int cardCount = player.getCardCount();
            GameConsole.write(name + ": " + cardCount + " " + getPluralCardLabel(cardCount));
        }

        GameConsole.writeLine();
    }

    private static String getPluralCardLabel(int count) {
        return count == 1 ? "card" : "cards";
    }
}
